errors/registry.go

```diff
@@ -0,0 +1,109 @@
+// Copyright © 2017 The Things Network
+// Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+package errors
+
+import (
+	"fmt"
+	"sync"
+)
+
+// registry represents an error type registry
+type registry struct {
+	sync.RWMutex
+	byCode map[Code]*ErrDescriptor
+}
+
+// Register registers a new error type
+func (r *registry) Register(err *ErrDescriptor) {
+	r.Lock()
+	defer r.Unlock()
+
+	if err.Code == 0 {
+		panic(fmt.Errorf("No code defined in error descriptor (message: `%s`)", err.MessageFormat))
+	}
+
+	if r.byCode[err.Code] != nil {
+		panic(fmt.Errorf("errors: Duplicate error code %v registered", err.Code))
+	}
+
+	err.registered = true
+	r.byCode[err.Code] = err
+}
+
+// Get returns the descriptor if it exists or nil otherwise
+func (r *registry) Get(code Code) *ErrDescriptor {
+	r.RLock()
+	defer r.RUnlock()
+	return r.byCode[code]
+}
+
+// reg is a global registry to be shared by packages
```

Comment:
A global registry is probably not a good idea. Let's assume we will use this `errors` package in the backend, console, integrations, etc. Then, we'd have to constantly sync on error codes to make sure we don't register overlapping codes in different namespaces.